fabtests/unit/getinfo_test.c

```diff
@@ -964,6 +964,249 @@ getinfo_test(caps, 4, "Test for capability bit regression",
 getinfo_test(caps, 5, "Test if either FI_LOCAL_COMM or FI_REMOTE_COMM is set",
 	     NULL, NULL, 0, hints, NULL, test_comm_caps, NULL, 0)
 
+static int test_valid_optional_cap_info(struct fi_info *info,
+					uint64_t info_caps,
+					uint64_t tx_caps,
+					uint64_t rx_caps)
+{
+	if ((info->caps & info_caps) != info_caps) {
+		print_incorrect_caps(info->fabric_attr->prov_name, "info",
+				     info_caps, info->caps);
+		return -FI_EINVAL;
+	}
+
+	if ((info->tx_attr->caps & tx_caps) != tx_caps) {
+		print_incorrect_caps(info->fabric_attr->prov_name, "tx",
+				     tx_caps, info->tx_attr->caps);
+		return -FI_EINVAL;
+	}
+
+	if ((info->rx_attr->caps & rx_caps) != rx_caps) {
+		print_incorrect_caps(info->fabric_attr->prov_name, "rx",
+				     rx_caps, info->rx_attr->caps);
+		return -FI_EINVAL;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int test_invalid_optional_cap(char *node, char *service, uint64_t flags,
+				     struct fi_info *hints,
+				     struct fi_info **info)
+{
+	uint64_t caps;
+	uint64_t domain_caps;
+	uint64_t tx_caps;
+	uint64_t rx_caps;
+	int ret;
+
+	/* Cache provider supported caps used for validation later. */
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo failed\n");
+		goto out;
+	}
+
+	caps = (*info)->caps;
+	domain_caps = (*info)->domain_attr->caps;
+	tx_caps = (*info)->tx_attr->caps;
+	rx_caps = (*info)->rx_attr->caps;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Feed in all unsupported capabilities as optional caps. These should
+	 * not be set in the returned fi_infos.
```

Comment:
This isn't quite true.  The provider could return multiple fi_info structures with different capabilities.  That's true today.  So, some caps which aren't set in the topmost fi_info may be set in later fi_info's.  You could probably just say "Many of the unsupported caps are highly likely to be unset in returned fi_infos".  I'm not sure there are any guarantees here without checking the full list of fi_info's, which doesn't seem worthwhile.  I think the structure of this test is reasonable, but the check for invalid optional_caps may need to be:

```
/* At least 1 optional cap should have been rejected */
if (!(~info->caps & hints->optional_caps))
    return -FI_EINVAL;
```
```

Reply:
Good point. I don't think it is worthwhile to iterate through the entire list. I'll update to the requested change.